structured/privilege.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package structured
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/privileges"
+)
+
+// rawToPrivilegeList converts a set of privileges
+// into a PrivilegeList.
+func rawToPrivilegeList(m uint32) privileges.PrivilegeList {
+	ret := privileges.PrivilegeList{}
+	for _, p := range privileges.SortedPrivileges {
+		if m&(1<<p) != 0 {
+			ret = append(ret, p)
+		}
+	}
+
+	return ret
+}
+
+// Sort methods for the PrivilegeDescriptor.Users list.
+type userPrivilegeList []*UserPrivileges
+
+func (upl userPrivilegeList) Len() int {
+	return len(upl)
+}
+
+func (upl userPrivilegeList) Swap(i, j int) {
+	upl[i], upl[j] = upl[j], upl[i]
+}
+
+func (upl userPrivilegeList) Less(i, j int) bool {
+	return upl[i].User < upl[j].User
+}
+
+// findUser looks for a specific user in the list.
+// Returns (nil, false) if not found, or (obj, true) if found.
+func (p *PrivilegeDescriptor) findUser(user string) (*UserPrivileges, bool) {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx < len(p.Users) && p.Users[idx].User == user {
+		return p.Users[idx], true
+	}
+	return nil, false
+}
+
+// findOrCreateUser looks for a specific user in the list, creating it if needed.
+func (p *PrivilegeDescriptor) findOrCreateUser(user string) *UserPrivileges {
+	idx := sort.Search(len(p.Users), func(i int) bool {
```

Comment:
this can't deal with a `nil` receiver, it seems. Just wondering because you call `GetDescriptors()` above (https://github.com/cockroachdb/cockroach/pull/2072/files#r36966395) which can give you a `nil`.